강원대 BE_손상욱 - Step3 위시리스트 구현

thstkddnr20 opened the pull request:
## 주요 변경 사항
1. 비밀번호를 그대로 저장하지 않고 암호화하여 DB에 저장합니다.
2. ArgumentResolver를 사용하여 LoginUser 어노테이션이 있는 파라미터에 자동으로 유저 정보를 주입합니다.
3. Jwt 필터에서 로그인 한 사용자의 역할만 보내주던 것에서 UserInfo 클래스를 사용하여 사용자의 역할, id를 함께 보내줍니다.
4. 위시리스트 API를 구현하고 e2e 테스트를 작성하였습니다.

## 맞닥뜨린 문제
1. 인증과 인가 처리에서의 문제점
사용자의 인증은 필터에서, 인가는 인터셉터에서 처리를 하고 있습니다. 여기서 추후 서비스가 복잡해진다고 가정할 때,

   예를 들어 관리자만 사용할 수 있는 API를 하나 추가해본다고 가정: 필터의 화이트리스트에 api path 등록 필요, AdminOnly 어노테이션을 핸들러의 메서드에 붙여주는 작업 필요

   이 2가지 작업이 모두 만족해야 관리자만 사용할 수 있는 API가 정상적으로 동작하게 됩니다.
-> 이는 프로그래머의 실수를 유발하기 쉬운 내용이라고 생각됩니다.

   또한 AdminOnly는 필터가 무조건 작동한다는 전제가 있어야 의미가 있는 어노테이션이기 때문에 이 둘의 관계가 사실 썩 유지보수하기 좋다고 생각되지는 않습니다.

   하지만 다른 좋은 방법이 떠오르지 않아 이대로 적용시켰습니다.....

2. 같은 경로에서의 필터 처리 문제
   예를 들어 상품 저장은 경로가 "[POST] /api/products"이고 **관리자 인가 검증**을 꼭 하는 부분이라고 가정하고, 상품 조회는 "[GET] /api/products"이며 **로그인 없이 사용 가능**한 API인 상황입니다.

   이때 물론 필터에서 HttpServletRequest로 getMethod()와 getRequestUri()로 메서드와 경로를 모두 검사하여 필터를 작동시킬지 말지 결정할 수 있겠지만, 유지보수하는 관점에서 생각해본다면 직접 필터의 화이트리스트 또는 블랙리스트에 메서드와 경로에 대해서 모두 작성해줘야하고 수정사항이 생길 시 해당 경로를 또 찾고 수정하고 하는 부분에서 좋지 않다고 생각합니다. 
-> 지금은 간단하여 물론 추가해줘도 되겠지만 나중에 API가 점점 많아진다면 큰 파장을 불러일으키지 않을까 싶습니다.

## 중점적으로 리뷰받고 싶은 곳
맞닥뜨린 문제에서 멘토님의 생각이 궁금합니다!

## 추가 내용
테스트가 종료될 때 truncate를 사용하였더니 외래키 제약 조건으로 인하여 사용할 수 없었고 delete from {table};로 테이블의 내용을 삭제해주었습니다. 

![image](https://github.com/user-attachments/assets/be5555b5-6d22-4905-98d7-5a368a923cdd)
맞닥뜨린 문제의 2번에서 생각해본 여러가지 내용을 같이 첨부합니다.

Commit:
docs: README.md step3 구현 기능 작성

Commit:
feat: 사용자의 패스워드를 암호화하여 저장

Commit:
feat: Wishlist db 세팅
cleanup.sql의 데이터 삭제를 truncate에서 delete from으로 변경 -> truncate는 제약조건을 고려하지 않는다

Commit:
fix: jwt 필터에서 사용자 정보 중 role만 반환하는 것에서 id까지 포함한 UserInfo를 반환

Commit:
feat: argumentResolver와 커스텀 어노테이션 LoginUser로 사용자의 정보를 가져옴

Commit:
feat: 위시리스트 등록 구현

Commit:
feat: 위시리스트 목록 조회 구현
join을 사용한 데이터를 불러오기 위해 repository가 dto 의존적이게 된 부분이 있다

Commit:
feat: 위시리스트 삭제 구현

Commit:
refactor: 위시리스트 예외 로직 추가

Commit:
fix: 이미 저장된 위시리스트 재등록 시 에러 처리

Commit:
test: 위시리스트 e2e 테스트 작성

Commit:
docs: README.md step3 구현 기능 추가 작성

Commit:
fix: 사용자 API에서는 이름을 검증하여 문의하도록 수정

leaf-upper commented on build.gradle:
```diff
@@ -22,6 +22,8 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-validation'
+    implementation 'org.springframework.security:spring-security-crypto:6.5.0'
```
스프링 시큐티리를 사용하셨군요~
아마 비밀번호 암호화만 하려고 해당 의존성을 추가하신것 같아요
여러가지 프레임워크를 익히고 사용해보는건 너무나 좋지만
현재 NextStep에서 추구하는 방향성과는 약간 다른것 같아요 ㅠ
좋은 프레임워크를 사용하여 애플리케이션을 개발하기 보다는! 현재는 조금더 기능요구사항에 집중해서 개발해보시면 좋을것 같습니다. 비밀번호 암호화는 아직까지는 기능요구사항이 아닌것 같아요~ 해당 의존성은 제거하고 PasswordEncoder또한 일단은 사용하지 않으셔도 좋을것 같습니다.

leaf-upper commented on src/main/java/gift/common/argumentResolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,27 @@
+public class LoginUserArgumentResolver implements HandlerMethodArgumentResolver {
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(LoginUser.class);
```
만약 LoginUser 애너테이션이 붙은 파라미터가 UserInfo가 아닌경우 (개발자의 실수) 어떻게 될까요?

thstkddnr20 replied:
UserInfo를 사용한다는 것을 명시하기 위해 파라미터의 타입을 검사하는 로직을 and 조건으로 같이 넣어줘야 합니다!

leaf-upper commented on src/main/java/gift/common/config/FilterConfig.java:
```diff
@@ -15,7 +15,8 @@ public class FilterConfig {
             "/admin/users/*",
             "/api/users/password",
-            "/api/users/admin"
+            "/api/users/admin",
```
같은 경로에서의 필터 처리 문제가 인증이 필요한곳에서만 인증 Filter를 넣기 때문인것 같네요~
저였다면 인증필터는 모든 Path에 대해서 태우고, Filter 레벨에서 Http Spec을 전부 확인하고 인증을 할것 같아요.

```
이때 물론 필터에서 HttpServletRequest로 getMethod()와 getRequestUri()로 메서드와 경로를 모두 검사하여 필터를 작동시킬지 말지 결정할 수 있겠지만, 유지보수하는 관점에서 생각해본다면 직접 필터의 화이트리스트 또는 블랙리스트에 메서드와 경로에 대해서 모두 작성해줘야하고 수정사항이 생길 시 해당 경로를 또 찾고 수정하고 하는 부분에서 좋지 않다고 생각합니다.
 ```

API를 만들때 특정 API를 어떤 사람이 사용할지 코드로 작성하는건 당연한 부분이라고 생각하는데요, 이런 부분은 수정이 필요한경우에 수정되어야 할것 같아요. 어떤 부분에서 좋지 않다고 생각하는지 조금만 더 생각하시는 부분을 알려주실 수 있으실까요?

leaf-upper replied:
```
인증과 인가 처리에서의 문제점
사용자의 인증은 필터에서, 인가는 인터셉터에서 처리를 하고 있습니다. 여기서 추후 서비스가 복잡해진다고 가정할 때,

예를 들어 관리자만 사용할 수 있는 API를 하나 추가해본다고 가정: 필터의 화이트리스트에 api path 등록 필요, AdminOnly 어노테이션을 핸들러의 메서드에 붙여주는 작업 필요

이 2가지 작업이 모두 만족해야 관리자만 사용할 수 있는 API가 정상적으로 동작하게 됩니다.
-> 이는 프로그래머의 실수를 유발하기 쉬운 내용이라고 생각됩니다.

또한 AdminOnly는 필터가 무조건 작동한다는 전제가 있어야 의미가 있는 어노테이션이기 때문에 이 둘의 관계가 사실 썩 유지보수하기 좋다고 생각되지는 않습니다.

하지만 다른 좋은 방법이 떠오르지 않아 이대로 적용시켰습니다.....
```

필터를 전체 Path에 대해서 적용해보는건 어떨까요?
Filter 자체에서 인증이 필요한 Path와 인증이 필요하지 않은 Path을 알고있고 적절하게 처리하면 되지않을까 싶습니다!

thstkddnr20 replied:
> API를 만들때 특정 API를 어떤 사람이 사용할지 코드로 작성하는건 당연한 부분이라고 생각하는데요, 이런 부분은 수정이 필요한경우에 수정되어야 할것 같아요. 어떤 부분에서 좋지 않다고 생각하는지 조금만 더 생각하시는 부분을 알려주실 수 있으실까요?

API가 추가되거나 수정되는 상황에서 **프로그래머가 직접** 해당 경로와 메서드를 직접 추가해줘야 한다
-> 이 프로그래머가 직접 한다는 사실 마음에 들지 않았습니다..

그래서 "직접 경로를 추가하거나 수정하지 않고 컴파일 시점에 핸들러의 커스텀 어노테이션을 검사하여 정보를 긁어와서 자동으로 추가해 줄 수 있지 않을까"를 생각했던 것 같습니다.....
제가 너무 어렵게 생각한 것 같습니다..

leaf-upper replied:
@thstkddnr20 
말씀해주신 방식도 충분히 가능해요~
Spring Bean에서 선언된 메서드들을 읽고, 
해당 메서드에 붙어있는 커스텀 애너테이션을 읽어서 허용하게 하든지, 인증하게 하든지 하는게 충분히 가능한 방식입니다.
물론 해당 방식의 코드를 작성하려면 조금 시간이 걸리긴 하겠지만 그래도 한번 작성해놓으면 잘사용해볼수 있을것 같아요.
좋은 고민입니다.

고민해보신 방식은 이런방식이면 어떨까요?

ex)
1. requestMappingHandlerMapping Bean을 읽는다. <- 이곳에서는 컨트롤러에 매핑되는 핸들러 메서드들이 등록됩니다.
2. 해당Bean에서 HandlerMethod()를 조회한다.
3. HandlerMethod의 커스텀 애너테이션을 조사한다.
4. 커스텀애너테이션에서 정보를 읽고 HandlerMethod와 결합하여 인증 및 인가 Path 정보를 Filter에 등록한다.

leaf-upper commented on src/main/java/gift/common/config/AdminUserInitializer.java:
```diff
@@ -17,6 +18,6 @@ public AdminUserInitializer(UserRepository userRepository) {
     @Override
     public void run(ApplicationArguments args) {
-        userRepository.save(new User("admin@admin.com", "1234", Role.ADMIN));
+        userRepository.save(new User("admin@admin.com", PasswordEncoder.encode("1234"), Role.ADMIN));
```
👍 👍 어드민 유저를 만드셨군요~ 좋은것 같아요! 고생하셨습니다

leaf-upper commented on src/main/java/gift/common/filter/JwtAuthenticationFilter.java:
```diff
@@ -41,6 +42,9 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
             }
 
             if (authHeader == null) {
+                if (request.getCookies().length == 0) {
```
Authorization헤더가 없으면 쿠키를 보나요?
쿠키에 Jwt를 담고 사용하는 느낌인가보네요~
Authorization 헤더만 봐도 되지않을까요?

thstkddnr20 replied:
이전에 구현한 관리자 화면(폼)에서 관리자임을 검사하기 위해 추가하였습니다
폼 화면과 api의 로그인처리를 동시에 하기 위해 이런 로직을 넣었습니다

leaf-upper commented on src/main/java/gift/repository/WishlistRepository.java:
```diff
@@ -0,0 +1,90 @@
+@Repository
+public class WishlistRepository {
+
+    private final JdbcTemplate jdbcTemplate;
```
👍 👍 Repository코드는 굉장히 깔끔하게 잘 작성해주셨네요 ㅎㅎ
고생하셨습니다!

leaf-upper reviewed (changes requested):
안녕하세요 상욱님~
인증인가 관련하여 몇가지 코멘트를 남겨드렸는데요, 
혹시 확인해보시고 궁금한점 있으시면 코멘트 남겨주세요!
Filter를 전체로 적용하고 Filter에서 Http Method 및 URI에 따라 인증 코드를 작성해보는건 어떨까요?
위 부분 적용해주시고 리뷰요청 주시면 다시 읽어볼게요!

Commit:
refactor: step3 피드백 반영하여 리팩토링
1. spring security 의존성 제거, PasswordEncoder 제거
2. 필터 화이트리스트 방식에서 블랙리스트 방식으로 변경, http method까지 같이 검사 진행
3. FilterRegistrationBean으로 필터 등록하는 것에서 Component + Order 방식으로 필터 생성 변경

Commit:
test: 요구사항 변경에 따른 상품 API 테스트 변경
로그인 없이 상품 조회만 가능, 그 외에는 모두 로그인 필요

leaf-upper commented on build.gradle:
```diff
@@ -22,7 +22,6 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-validation'
-    implementation 'org.springframework.security:spring-security-crypto:6.5.0'
```
👍 👍 Spring Security는 조금더 이후에 프로젝트를 할때 사용해보면 좋을것 같아요!
고생하셨습니다 ㅎㅎ

leaf-upper commented on src/main/java/gift/common/argumentResolver/LoginUserArgumentResolver.java:
```diff
@@ -12,7 +12,7 @@
     @Override
     public boolean supportsParameter(MethodParameter parameter) {
-        return parameter.hasParameterAnnotation(LoginUser.class);
+        return parameter.hasParameterAnnotation(LoginUser.class) && parameter.getParameterType().equals(UserInfo.class);
```
👍 👍

leaf-upper reviewed (approved):
👍 👍 고생하셨습니다 상욱님
남겨드렸던 코멘트들을 잘반영해주셔서 어프로브 드릴게요~
위시리스트 구현 고생 많으셨어요~!
